File: code/Project_4/BT5/python/prolems/problem_1_4.py
class Graph:
    def __init__(self, n):
        self.n = n
        self.matrix = [[0] * n for _ in range(n)]
    
    def add_edge(self, v, w):
        if 0 <= v < self.n and 0 <= w < self.n:
            self.matrix[v][w] = 1
    
    def del_edge(self, v, w):
        if 0 <= v < self.n and 0 <= w < self.n:
            self.matrix[v][w] = 0
    
    def edges(self):
        result = []
        for i in range(self.n):
            for j in range(self.n):
                if self.matrix[i][j]:
                    result.append((i, j))
        return result
    
    def incoming(self, v):
        if not (0 <= v < self.n):
            return []
        result = []
        for i in range(self.n):
            if self.matrix[i][v]:
                result.append(i)
        return result
    
    def outgoing(self, v):
        if not (0 <= v < self.n):
            return []
        result = []
        for j in range(self.n):
            if self.matrix[v][j]:
                result.append(j)
        return result
    
    def source(self, v, w):
        if 0 <= v < self.n and 0 <= w < self.n:
            return self.matrix[v][w] != 0
        return False
    
    def target(self, v, w):
        if 0 <= v < self.n and 0 <= w < self.n:
            return self.matrix[v][w] != 0
        return False
    
    def display(self):
        for row in self.matrix:
            print(row)

def main():
    # Test với đồ thị 3 đỉnh
    g = Graph(3)
    g.add_edge(0, 1)
    g.add_edge(1, 2)
    g.add_edge(0, 2)
    
    print("Graph matrix:")
    g.display()
    
    print(f"\nEdges: {g.edges()}")
    print(f"Incoming to 2: {g.incoming(2)}")
    print(f"Outgoing from 0: {g.outgoing(0)}")
    print(f"Is 0 connected to 1? {g.source(0, 1)}")
    print(f"Is 1 target of 0? {g.target(0, 1)}")

if __name__ == "__main__":
    main()

File: code/Project_4/BT5/c++/problems/problem_1_4.cpp
#include <iostream>
#include <vector>
using namespace std;

class Graph {
private:
    int n;
    vector<vector<int>> matrix;

public:
    Graph(int vertices) : n(vertices) {
        matrix.resize(n, vector<int>(n, 0));
    }
    
    void add_edge(int v, int w) {
        if (v >= 0 && v < n && w >= 0 && w < n) {
            matrix[v][w] = 1;
        }
    }
    
    void del_edge(int v, int w) {
        if (v >= 0 && v < n && w >= 0 && w < n) {
            matrix[v][w] = 0;
        }
    }
    
    vector<pair<int, int>> edges() {
        vector<pair<int, int>> result;
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                if (matrix[i][j]) {
                    result.push_back({i, j});
                }
            }
        }
        return result;
    }
    
    vector<int> incoming(int v) {
        vector<int> result;
        if (v < 0 || v >= n) return result;
        for (int i = 0; i < n; i++) {
            if (matrix[i][v]) {
                result.push_back(i);
            }
        }
        return result;
    }
    
    vector<int> outgoing(int v) {
        vector<int> result;
        if (v < 0 || v >= n) return result;
        for (int j = 0; j < n; j++) {
            if (matrix[v][j]) {
                result.push_back(j);
            }
        }
        return result;
    }
    
    bool source(int v, int w) {
        if (v >= 0 && v < n && w >= 0 && w < n) {
            return matrix[v][w] != 0;
        }
        return false;
    }
    
    bool target(int v, int w) {
        if (v >= 0 && v < n && w >= 0 && w < n) {
            return matrix[v][w] != 0;
        }
        return false;
    }
    
    void display() {
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                cout << matrix[i][j] << " ";
            }
            cout << endl;
        }
    }
};

int main() {
    Graph g(3);
    g.add_edge(0, 1);
    g.add_edge(1, 2);
    g.add_edge(0, 2);
    
    cout << "Graph matrix:" << endl;
    g.display();
    
    auto edge_list = g.edges();
    cout << "\nEdges: ";
    for (auto edge : edge_list) {
        cout << "(" << edge.first << "," << edge.second << ") ";
    }
    cout << endl;
    
    auto incoming = g.incoming(2);
    cout << "Incoming to 2: ";
    for (int v : incoming) {
        cout << v << " ";
    }
    cout << endl;
    
    auto outgoing = g.outgoing(0);
    cout << "Outgoing from 0: ";
    for (int v : outgoing) {
        cout << v << " ";
    }
    cout << endl;
    
    cout << "Is 0 connected to 1? " << (g.source(0, 1) ? "Yes" : "No") << endl;
    cout << "Is 1 target of 0? " << (g.target(0, 1) ? "Yes" : "No") << endl;
    
    return 0;
}